tests/js/legendTests.js

```diff
@@ -0,0 +1,205 @@
+        label: "Two"
+    }, {
+        id: "id2",
+        value: 20,
+        label: "Three"
+    }, {
+        id: "id3",
+        value: 45,
+        label: "Four"
+    },
+    {
+        id: "id4",
+        value: 26,
+        label: "Five"
+    }
+  ];
+
+  gpii.tests.chartAuthoring.objectArrayRemove = [{
+      id: "id0",
+      value: 15,
+      label: "One"
+  }, {
+      id: "id1",
+      value: 67,
+      label: "Two"
+  }, {
+      id: "id2",
+      value: 20,
+      label: "Three"
+  }
+];
+
+    gpii.tests.chartAuthoring.mouseOverListener = function (data, i, that) {
+        that.mouseOverListenerCalled = true;
+    };
+
+    // convenience function for easing testing of colors (jquery returns only RGB)
+    // Based off http://stackoverflow.com/questions/4262417/jquery-hex-to-rgb-calculation-different-between-browsers
+    gpii.tests.chartAuthoring.hexToRGB = function(hexStr){
+        // note: hexStr should be #rrggbb
+        var hex = parseInt(hexStr.substring(1), 16);
+        var r = (hex & 0xff0000) >> 16;
+        var g = (hex & 0x00ff00) >> 8;
+        var b = hex & 0x0000ff;
+        return "rgb("+r+", "+g + ", " +b + ")";
+    }
+
+    gpii.tests.chartAuthoring.testMouseOverListener = function(that) {
+      jqUnit.assertFalse("The mouseover listener for legend rows has not been triggered", that.mouseOverListenerCalled);
+      var oneD3Row = that.jQueryToD3($(that.locate("row")[0]));
+      oneD3Row.on("mouseover")();
+      jqUnit.assertTrue("The mouseover listener for legend rows has been triggered", that.mouseOverListenerCalled);
+    }
+
+    gpii.tests.chartAuthoring.validateLegend = function (that) {
+        var table = that.locate("table");
+
+        // Test the legend creation
+        jqUnit.assertNotEquals("The TABLE element for the legend is created with the proper selector", 0, table.length);
+
+        jqUnit.assertEquals("A TR element has been created for each value in the dataset, with proper selectors", that.model.dataSet.length, that.locate("row").length);
+
+        var d3ColorCells = that.jQueryToD3($(that.locate("legendColorCell")));
+        d3ColorCells.each(function (d, i) {
+
+            jqUnit.assertEquals("The data colors are filled correctly in the legend", gpii.tests.chartAuthoring.hexToRGB(d.color), ($(this).css("background-color")));
+        });
+
+        var d3LabelCells = that.jQueryToD3($(that.locate("labelCell")));
+        d3LabelCells.each(function (d, i) {
+            jqUnit.assertEquals("The data labels are applied correctly in the legend", d.label, ($(this).html()));
+        });
+
+        var d3LabelCells = that.jQueryToD3($(that.locate("valueCell")));
+        d3LabelCells.each(function (d, i) {
+            jqUnit.assertEquals("The data values are applied correctly in the legend", d.value, ($(this).html()));
+        });
+
+    };
+
+    jqUnit.test("Test the legend component created based off an array of objects, unsorted", function () {
+        jqUnit.expect(44);
```

Comment:
The functional flow for testing sorted and unsorted legends are same. Can we consolidate into a function to reuse.